Marketplace: Seller permissions, Allow guests to become sellers, Create a shop, Post products and Edit their shop settings

lcampanis opened the pull request:
Clean copy of: https://github.com/reactioncommerce/reaction/pull/1591 https://github.com/reactioncommerce/reaction/pull/1669

**This branch is ready for review.**

- Allow Sellers to add products via `Add Product`to their shop only, when they have the `createProduct` role
- Amended `shopIdAutoValue` to check loggedIn user has a shop, or fallback on parent shop
- Amended `getCurrentShop` as a start towards shop switching UI
- Amended `Product` publication to allow sub-sellers (admin) to add products to their shop
- Amended `gridControls` so that sub-sellers with `createProduct` permission can edit their products from the grid (create a new grid view for this to work)
- Allow guests to become sellers when marketplace module is enabled, `defaultSellerRoles` in Account Schema
- Added Become a seller panel in user profile page only when Marketplace plugin is enabled and `AllowGuestSeller` option is enabled
- Amended `shop/createShop` to allow guests to create a shop only once, when marketplace is enabled
- New meteor method `shop/getSellerShopId` to safely get seller shop Id, or default to parent shopId
- New `lib/api/core` for `Reaction` 
- New core functions `Reaction.addRolesToVisitors` and `Reaction.isPackageEnabled`
- Fixed `Reaction.getPackageSettings`
- Marketplace dashboard settings under `Catalog` settings with option to `allowGuestSellers` (might be better under `Settings`?)
- Fix loggedIn User dropdown perms - PR https://github.com/reactioncommerce/reaction/pull/1639
- Allow Sellers to edit their `shop settings` and change the shop's `Base Currency` https://github.com/reactioncommerce/reaction/issues/1711

### In Progress
1. Amend `Products` Grid View and Publications to show "published" products from multiple sellers, with permissions for logged in seller to edit directly from the grid

### To Do
1. Product cart flow to allow products from multiple sellers **TBD**
1. Payment flow to allow payments to multiple sellers **TBD**
1. Allow Sellers to edit their shop URL by updating Shop.domains (domain, subdomain, path based URLs)
1. Allow Tag management per shop 
1. Expand on dashboard settings, with `enableBanAccounts` see https://github.com/reactioncommerce/reaction/issues/1578, etc)

### Notes
- Introduced new client-only `SellerShops` collection - This leaves the Shops collection intact with the parent shop

lcampanis commented:
@aaronjudd Okay checking

aaronjudd commented:
@lcampanis this is really fantastic! 

@mikemurray fixed the product grid so that we could test the other day, you might want to review those changes but I think they made sense.

@jshimko @zenweasel @mikemurray some testing suggestions:

- Login as initial owner, verify marketplace is enabled, go to accounts, invite a user.
- Incognito-> Login as this new invited user.  
- View profile, verify "become a seller" button.  Should go away if marketplace is disabled.
- Create a new shop profile as the user
- Create product, etc.. 
- verify different account permissions, admin lets you change settings, seller only gives you your own settings.
- order processing need some testing, not exactly sure but seems like seller should see his, admin sees all?

Note permissions in admin are not always correct(you can see the header/panels/toggle, but not settings). Likely related to #797, maybe more.. we should creating individual tickets for fixing any of these permissions slips.

@lcampanis we're going to need some help documentation all this.  maybe you can throw in some additional testing instructions.

This is just my first pass through the UI, so I might not have caught all the changes.. code review next.

aaronjudd commented on client/modules/core/main.js:
```diff
@@ -196,6 +196,17 @@ export default {
+    }
+
+    return this.getShopId();
+  },
```
this means for now, there can only be one shop per user, which I think a good start ( just something to document.  I can easily see having multiple shops per user, as long as we keep getShopId() as the source of truth that should be easy enough to do.

aaronjudd commented on imports/plugins/included/marketplace/client/templates/settings/sellerShopSettings.js:
```diff
@@ -0,0 +1,153 @@
+          Meteor.call("shop/updateBrandAssets", asset, (error, result) => {
+            if (error) {
+              // Display Error
+              return Alerts.toast("Couldn't update brand asset.", "error");
```
Alerts should also have i18n translations.

lcampanis replied:
Fixed

lcampanis replied:
Fixed in core as well

aaronjudd commented on private/data/i18n/en.json:
```diff
@@ -642,6 +641,13 @@
           "orderTotal": "Total"
         }
       },
+      "marketplace": {
```
Marketplace specific translations could be moved to `marketplace/server/i18n`.   See the payment or taxes plugins for examples of loading translations from there.  Just adding the `en.json` will take care of it, the LingoHub service will translate it into individual language files, but we'll need to manually update the imports once that is done.

lcampanis replied:
Fixed

lcampanis commented on client/modules/accounts/templates/profile/profile.js:
```diff
@@ -1,6 +1,7 @@
 import { Meteor } from "meteor/meteor";
 import { Template } from "meteor/templating";
+import * as Collections from "/lib/collections";
+import { Marketplace } from "/imports/plugins/included/marketplace/lib/api";
```
Agreed it's better to keep them separate, this is why a `Reaction.Marketplace` solution for methods is also problematic in some ways. Some functions tied around the marketplace need to live in Reaction, at least the core ones that check sellerId, sellerShop and the core functionality that checks if the module is enabled and the user is a guest, very simple functions. They're not heavy and probably they can be consolidated with existing functions, moving forward.
Should we place it along with the other two functions in core.jshttps://github.com/reactioncommerce/reaction/pull/1758/files/9f64febe5e71033b5691ed92946ed52c7c3db53f#diff-541cd596a355327d6884076c3ee12121?
Although the current /import solution is clean, it is not practical. Whereas putting marketplace core functions in core.js to check shopIds and permissions when the module is enabled is probably less clean but more practical. This way the module can be deleted and disabled.

ltouro commented on imports/plugins/core/dashboard/client/components/toolbar.js:
```diff
@@ -80,7 +78,7 @@ class PublishControls extends Component {
         onClick={() => {
           Reaction.showActionView({
-            i18nKeyTitle: "dashboard.coreTitle",
+            i18nKeyTite: "dashboard.coreTitle",
```
there is a typo in i18nKey**Tite**

lcampanis replied:
Fixed

zenweasel replied:
I don't see this being fixed?
https://github.com/reactioncommerce/reaction/blob/dc2703a93df5d231e432f014de965e9518876467/imports/plugins/core/dashboard/client/components/toolbar.js#L81-L81

lcampanis commented on client/modules/core/helpers/apps.js:
```diff
@@ -125,11 +125,8 @@ export function Apps(optionHash) {
-
-  return sortedApps;
+  // sort cycle to ensure order aka. is registry.priority working? .sort((a, b) => a.priority - b.priority).slice();
+  return reactionApps;
```
@aaronjudd Do you need us to change/revert something here?

zenweasel replied:
I believe that change needs to get reverted so we are returning `sortedApps`

lcampanis commented:
@aaronjudd All changes are now done and tested. Release 19 merged as well. A massive step forward! 
Not sure if some of the questions were for us. Any other changes though, let us know..
If it's there's no objections we'll go ahead and create a marketplace publication to show all published seller items in the new grid.

jshimko commented:
@lcampanis Is there supposed to be a way for a marketplace user to specify their shop when managing products?  When I invite a user, log in as that user, click "become a marketplace seller" in their profile, then create a product, it adds it to the original shop.  I don't see any way to specify otherwise or even view the secondary shop.  Am I missing something?

zenweasel commented:
I think it's a little confusing to have an on-off toggle that says "Marketplace settings"

lcampanis commented on imports/plugins/core/dashboard/client/components/toolbar.js:
```diff
@@ -36,7 +36,7 @@ class PublishControls extends Component {
     if (typeof this.props.onViewContextChange === "function") {
-      this.props.onViewContextChange(event, isChecked ? "administrator" : "customer");
+      this.props.onViewContextChange(event, isChecked ? "customer" : "administrator");
     }
```
And I was wondering where that came from @aaronjudd 
![image](https://cloud.githubusercontent.com/assets/670954/22725287/2e6091c0-edd5-11e6-8985-896635fbbea1.png)

lcampanis commented on imports/plugins/core/revisions/client/components/publishControls.js:
```diff
@@ -337,7 +337,7 @@ class PublishControls extends Component {
         onClick={() => {
           Reaction.showActionView({
-            i18nKeyTite: "dashboard.coreTitle",
+            i18nKeyTitle: "dashboard.coreTitle",
```
Looks like there are some
![image](https://cloud.githubusercontent.com/assets/670954/22725365/a98c4fe2-edd5-11e6-98ff-a2b3a5c068a8.png)

zenweasel replied:
Yeah, those are usages. But I don't see the actual translation

lcampanis commented on imports/plugins/core/ui-tagnav/client/components/tagNav/tagNav.js:
```diff
@@ -270,7 +249,7 @@ Template.tagNav.helpers({
 
   canEdit() {
-    return Template.instance().data.editable && Reaction.isPreview() === false;
+    return Template.instance().data.editable;
```
@aaronjudd ?
![image](https://cloud.githubusercontent.com/assets/670954/22725401/df350c6a-edd5-11e6-92a4-362ff671cdaf.png)

aaronjudd replied:
@mikemurray I believe these were removed as well...

lcampanis commented on imports/plugins/included/product-detail-simple/client/containers/variantListContainer.js:
```diff
@@ -195,7 +195,7 @@ function composer(props, onData) {
   let editable;
 
-  if (Reaction.isPreview() === true) {
+  if (Reaction.Router.getQueryParam("as") === "customer") {
```
Aaron's again

aaronjudd replied:
@mikemurray I believe these were removed as well...

lcampanis commented on lib/api/core.js:
```diff
@@ -0,0 +1,55 @@
+import { Meteor } from "meteor/meteor";
+import { Shops } from "/lib/collections";
+import { SellerShops } from "/imports/plugins/included/marketplace/lib/collections";
```
Yes. I think we'll add that check function in here directly and remove these imports

lcampanis commented on lib/api/core.js:
```diff
@@ -0,0 +1,55 @@
+    }
+  }
+
+  return Reaction.getShopId();
```
No issue I can think of. You won't be able to call this function before Reaction is init-ed as it has to be extended first, and then exported

lcampanis commented:
@aaronjudd All fixes/changes from the review are now done. I hope we didn't miss anything.
@zenweasel regarding dashboard translations, they don't work for owner either, so most probably a problem with core translations when imported from other plugins?
@jshimko Nothing to set or choose, you can login as a guest, got to profile and click become a seller, then publishing a product should be visible in mongo with the new shopId. 
We removed the /imports/.../marketplace statements and moved the function in core.js. 
Any more changes let us know..

lcampanis commented:
@aaronjudd Forgot to mention I'll write the gherkin tests this week ;)

aaronjudd commented:
👍 I was going to suggest that we need to look at additional tests as well... (mocha/chai integration..)

lcampanis commented:
Hi @mikemurray.  Thanks for reviewing. Some answers for you

> Created a product as a seller. Product does not show up in grid for either seller or shop owner. Sometimes it shows up, but If I make it visible it disappears after publish. Guess this is in #1822

Yes it is in #1822. Some core UI issues are still happening there, but we're close..


> Should I be able to edit the tags in the nav bar? Currently I'm able to enter edit mode and try to edit, but the update fails with an error.
> Cannot add tags to the product I created. Tag form is blank in side panel.
> Cannot checkout with my product. Shipping options don't appear, even though they're configured. (Using flat rate shipping options)

The new PR should fix permissions around tags (but not tested yet). 
We'll be creating new PRs for tags, cart and shipping, this week and the next.
 
> Am I supposed to be able to access my own catalog via URL? If so, can we put a link somewhere?

No seller homepage just yet. Should have one by next week 👍
